File: src/main/java/nucleon/event/FilterChainManager.java
package nucleon.event;

import nucleon.util.Args;

import java.lang.reflect.Modifier;
import java.util.HashMap;
import java.util.Map;

public final class FilterChainManager {

    public static final FilterChainManager GLOBAL = new FilterChainManager();

    @SuppressWarnings("unchecked")
    private static Class<? extends Event> resolveNearestHandleableParent(Class<? extends Event> eventClass) {
        Class<?> superclass = eventClass.getSuperclass();

        while (superclass != Event.class) {
            Class<? extends Event> eventSuperclass = (Class<? extends Event>) superclass;
            if (isEventHandleable(eventSuperclass)) {
                return eventSuperclass;
            }
            superclass = eventSuperclass.getSuperclass();
        }

        return null;
    }

    private static boolean isEventHandleable(Class<? extends Event> eventClass) {
        return !Modifier.isAbstract(eventClass.getModifiers()) || eventClass.isAnnotationPresent(AllowHandle.class);
    }

    private final Map<Class<? extends Event>, FilterChain<?>> event2chainMap = new HashMap<>();

    @SuppressWarnings("unchecked")
    public <T extends Event> FilterChain<T> acquireChainFor(Class<? extends T> eventClass) {
        Args.notNull(eventClass, "Event class");

        if (event2chainMap.containsKey(eventClass)) {
            return (FilterChain<T>) event2chainMap.get(eventClass);
        }

        Class<? extends Event> superclass = resolveNearestHandleableParent(eventClass);
        FilterChain<? super T> parentChain;

        if (superclass == null) {
            parentChain = null;
        } else {
            parentChain = acquireChainFor(superclass);
        }

        FilterChain<T> list = new FilterChain<>(eventClass, parentChain);
        event2chainMap.put(eventClass, list);
        return list;
    }
}

File: src/main/java/nucleon/event/FilterChainManager.kt
package nucleon.event

import java.lang.reflect.Modifier

object FilterChainManager {

    private fun isHandleable(type: Class<out Event>) =
        !Modifier.isAbstract(type.modifiers) || type.isAnnotationPresent(AllowHandle::class.java)

    @Suppress("UNCHECKED_CAST")
    private fun resolveNearestHandleableParent(baseType: Class<out Event>): Class<out Event>? {
        var superType = baseType.superclass
        while (superType != Event::class.java) {
            if (isHandleable(superType as Class<out Event>)) {
                return superType
            }
            superType = superType.superclass
        }
        return null
    }

    private val type2chainMap: MutableMap<Class<out Event>, FilterChain<*>> = HashMap()

    @Suppress("UNCHECKED_CAST")
    fun <T : Event> acquireChainFor(eventType: Class<out T>): FilterChain<T> {
        if (type2chainMap.contains(eventType)) {
            return type2chainMap[eventType] as FilterChain<T>
        }
        val parent = resolveNearestHandleableParent(eventType)
        val chain = FilterChain(eventType, if (parent != null) acquireChainFor(parent) else null)
        type2chainMap[eventType] = chain
        return chain
    }
}